File: pkg/R/MovingAverages.R
#-------------------------------------------------------------------------#
# TTR, copyright (C) Joshua M. Ulrich, 2007                               #
# Distributed under GNU GPL version 3                                     #
#-------------------------------------------------------------------------#

"SMA" <-
function(x, n=10) {

  # Simple Moving Average

  # http://www.fmlabs.com/reference/SimpleMA.htm
  # http://www.equis.com/Customer/Resources/TAAZ/Default.aspx?c=3&p=74
  # http://linnsoft.com/tour/techind/movAvg.htm
  # http://stockcharts.com/education/IndicatorAnalysis/indic_movingAvg.html

  ma <- runMean( x, n )

  return(ma)
}

#-------------------------------------------------------------------------#

"EMA" <-
function (x, n=10, wilder=FALSE, ratio=NULL) {

  # Exponential Moving Average

  # http://www.fmlabs.com/reference/ExpMA.htm
  # http://www.equis.com/Customer/Resources/TAAZ/Default.aspx?c=3&p=74
  # http://linnsoft.com/tour/techind/movAvg.htm
  # http://stockcharts.com/education/IndicatorAnalysis/indic_movingAvg.html

  x <- try.xts(x, error=FALSE)
  
  # Count NAs, ensure they're only at beginning of data, then remove.
  # Avoid na.omit() because it will cause problems for reclass()
  x.na <- naCheck(x, n)

  # Initialize ma vector
  ma <- rep(1, NROW(x))
  ma[x.na$beg] <- mean(x[x.na$nonNA[1]:x.na$beg])

  # Determine decay ratio
  if(is.null(ratio)) {
    if(wilder) ratio <- 1/n
    else       ratio <- 2/(n+1)
  }

  # Call Fortran routine
  ma <- .Fortran( "ema", ia = as.double(x[x.na$nonNA]),
                         lia = as.integer(NROW(x.na$nonNA)),
                         n = as.integer(n),
                         oa = as.double(ma[x.na$nonNA]),
                         loa = as.integer(NROW(x.na$nonNA)),
                         ratio = as.double(ratio),
                         PACKAGE = "TTR" )$oa

  # Replace 1:(n-1) with NAs and prepend NAs from original data
  is.na(ma) <- c(1:(n-1))
  ma <- c( rep( NA, x.na$NAs ), ma ) 
  
  # Convert back to original class
  reclass(ma, x)
}

#-------------------------------------------------------------------------#

"DEMA" <-
function(x, n=10) {

  # Double Exponential Moving Average

  # http://www.fmlabs.com/reference/DEMA.htm

  dema <- 2 * EMA(x,n) - EMA(EMA(x,n),n)

  return( dema )
}

#-------------------------------------------------------------------------#

"WMA" <-
function(x, n=10, wts=1:n) {

  # Weighted Moving Average

  # http://www.fmlabs.com/reference/WeightedMA.htm
  # http://www.equis.com/Customer/Resources/TAAZ/Default.aspx?c=3&p=74
  # http://linnsoft.com/tour/techind/movAvg.htm

  if( !any( NROW(wts) == c( NROW(x), n ) ) )
    stop("Length of 'wts' must equal the length of 'x' or 'n'")
  if( n < 1 || n > NROW(x) )
    stop("Invalid 'n'")

  # Count NAs, ensure they're only at beginning of data, then remove.
  NAx <- sum( is.na(x) )
  NAw <- sum( is.na(wts) )
  NAs <- max( NAx, NAw )
  if( NAs > 0 ) {
    if( any( is.na(  x[-(1:NAx)]) ) )
      stop("'x' contains non-leading NAs")
    if( any( is.na(wts[-(1:NAw)]) ) )
      stop("'wts' contains non-leading NAs")
  }
  
  if( NROW(wts) == n ) {
    
    x <- na.omit(x)
    NAs <- NAx

    if( any(is.na(wts)) )
      stop("'wts' vector of length 'n' cannot have NA values")

    # Call Fortran routine
    ma <- .Fortran( "wma", ia = as.double(x),
                           lia = as.integer(NROW(x)),
                           wts = as.double(wts),
                           n = as.integer(n),
                           oa = as.double(x),
                           loa = as.integer(NROW(x)),
                           PACKAGE = "TTR" )$oa
   
  } else {
    
    xw <- na.omit( cbind(x, wts) )
    ma <- runSum( xw[,1]*xw[,2], n) / runSum(xw[,2], n)
  }

  # replace 1:(n-1) with NAs and prepend NAs from original data
  ma[1:(n-1)] <- NA
  ma <- c( rep( NA, NAs ), ma )

  return( ma )
}

#-------------------------------------------------------------------------#

"EVWMA" <-
function(price, volume, n=10) {

  # Elastic, Volume-Weighted Moving Average

  # http://linnsoft.com/tour/techind/evwma.htm

  if( !any( NROW(volume) == c( NROW(price), 1 ) ) )
    stop("Length of 'volume' must equal 1 or the length of 'price'")
  if( n < 1 || n > NROW(price) )
    stop("Invalid 'n'")

  # Count NAs, ensure they're only at beginning of data, then remove.
  NAp <- sum( is.na(price) )
  NAv <- sum( is.na(volume) )
  NAs <- max( NAp, NAv )
  if( NAs > 0 ) {
    if( any( is.na( price[-(1:NAp)]) ) )
      stop("'price' contains non-leading NAs")
    if( any( is.na(volume[-(1:NAv)]) ) )
      stop("'volume' contains non-leading NAs")
  }
  pv <- na.omit( cbind(price, volume) )

  # Initialize ma vector 
  ma <- rep(0, NROW(pv))
  ma[n] <- pv[n,1]

  if(NROW(volume)==1) {
    vSum <- rep(volume, NROW(pv))
  } else {
    vSum <- runSum(pv[,2], n)
    vSum[1:(n-1)] <- pv[1:(n-1),2]
  }

  # Call Fortran routine
  ma <- .Fortran( "evwma", ip = as.double(pv[,1]),
                           iv = as.double(pv[,2]),
                           ivs = as.double(vSum),
                           lia = as.integer(NROW(pv)),
                           n = as.integer(n),
                           oa = as.double(ma),
                           loa = as.integer(NROW(ma)),
                           PACKAGE = "TTR" )$oa

  # replace 1:(n-1) with NAs and prepend NAs from original data
  ma[1:(n-1)] <- NA
  ma <- c( rep( NA, max( NAp, NAv ) ), ma )

  return( ma )
}

#-------------------------------------------------------------------------#

"ZLEMA" <-
function (x, n=10, ratio=NULL) {

  # Zero-Lag Exponential Moving Average

  # http://www.fmlabs.com/reference/ZeroLagExpMA.htm
  # http://linnsoft.com/tour/techind/movAvg.htm

  x   <- as.vector(x)
  
  # Count NAs, ensure they're only at beginning of data, then remove.
  NAs <- sum( is.na(x) )
  if( NAs > 0 ) {
    if( any( is.na(x[-(1:NAs)]) ) )
      stop("Series contains non-leading NAs")
  }
  x   <- na.omit(x)

  # Initialize ma vector
  ma <- rep(1, NROW(x))
  ma[n] <- mean(x[1:n])

  # Determine decay ratio
  if(is.null(ratio)) {
    ratio <- 2/(n-1)
  }

  # Call Fortran routine
  ma <- .Fortran( "zlema", ia = as.double(x),
                           lia = as.integer(NROW(x)),
                           n = as.integer(n),
                           oa = as.double(ma),
                           loa = as.integer(NROW(ma)),
                           ratio = as.double(ratio),
                           PACKAGE = "TTR" )$oa

  # replace 1:(n-1) with NAs and prepend NAs from original data
  ma[1:(n-1)] <- NA
  ma <- c( rep( NA, NAs ), ma ) 
  
  return(ma)
}

File: src/movingAverages.f
c
c   TTR: Technical Trading Rules
c
c   Copyright (C) 2007-2008  Joshua M. Ulrich
c
c   This program is free software: you can redistribute it and/or modify
c   it under the terms of the GNU General Public License as published by
c   the Free Software Foundation, either version 3 of the License, or
c   (at your option) any later version.
c
c   This program is distributed in the hope that it will be useful,
c   but WITHOUT ANY WARRANTY; without even the implied warranty of
c   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
c   GNU General Public License for more details.
c
c   You should have received a copy of the GNU General Public License
c   along with this program.  If not, see <http://www.gnu.org/licenses/>.
c

c     Calculate Simple Moving Average (SMA)
c
c http://www.fmlabs.com/reference/SimpleMA.htm
c http://www.equis.com/Customer/Resources/TAAZ/Default.aspx?c=3&p=74
c http://linnsoft.com/tour/techind/movAvg.htm
c http://stockcharts.com/education/IndicatorAnalysis/indic_movingAvg.html
c 
c     http://en.wikipedia.org/wiki/Moving_average
c     I think Wikipedia is wrong. The MA shouldn't have
c     anything to do with _tomorrow's_ value.
c
c      subroutine sma(ia, lia, n, oa, loa)
c
c      integer lia, n, loa, i
c      double precision ia(lia), oa(loa)
c
c      do 10 i=n+1,lia
c
c        oa(i) = oa(i-1) + (ia(i) - ia(i-n))/n
c      
c   10 continue
c      end

c-----------------------------------------------------------------------c
c
c     Calculate Exponential Moving Average (EMA)
c
c http://www.fmlabs.com/reference/ExpMA.htm
c http://www.equis.com/Customer/Resources/TAAZ/Default.aspx?c=3&p=74
c http://linnsoft.com/tour/techind/movAvg.htm
c http://stockcharts.com/education/IndicatorAnalysis/indic_movingAvg.html
c
c     Thanks to Jeff Ryan for the Fortran implementation,
c     and more importantly, the encouragement!
c
c     ia    : input array
c     lia   : length of input array
c     n     : number of periods
c     oa    : output array
c     loa   : length of output array
c     ratio : weighting/decay ratio
c
      subroutine ema(ia, lia, n, oa, loa, ratio)

      integer lia, n, loa, i
      double precision ia(lia), oa(loa)
      double precision ratio

      do 10 i=n+1,lia

        oa(i) = ia(i) * ratio + oa(i-1) * (1-ratio)

   10 continue
      end

c-----------------------------------------------------------------------c
c
c     Calculate Weighted Moving Average (WMA)
c
c http://www.fmlabs.com/reference/WeightedMA.htm
c http://www.equis.com/Customer/Resources/TAAZ/Default.aspx?c=3&p=74
c http://linnsoft.com/tour/techind/movAvg.htm
c
c     Only for the case of wts(n). If wts(lia), the
c     WMA can be calculated with 'runSP' ('runSumProd' in R).
c
c     ia   : input array
c     lia  : length of input array
c     wts  : weight array
c     n    : number of periods (= length of weight array)
c     oa   : output array
c     loa  : length of output array
c     num  : window numerator
c     den  : window denominator
c
      subroutine wma(ia, lia, wts, n, oa, loa)

      integer lia, n, loa, i, j
      double precision ia(lia), wts(n), oa(loa)
      double precision num, den
      
      do 10 i=n,lia

         num = 0.0D0
         den = 0.0D0

        do 20 j=i-n+1,i
          
          num = num + ( ia(j) * wts(n-i+j) )
          den = den + wts(n-i+j)

   20 continue

        oa(i) = num / den

   10 continue
      end

c-----------------------------------------------------------------------c
c
c     Calculate Elastic, Volume-Weighted Moving Average (EVWMA)
c
c http://linnsoft.com/tour/techind/evwma.htm
c  
c     ip    : input array, price
c     iv    : input array, volume
c     ivs   : input array, volume sum
c     lia   : length of input arrays
c     n     : number of periods
c     oa    : output array
c     loa   : length of output array
c
      subroutine evwma(ip, iv, ivs, lia, n, oa, loa)

      integer lia, n, loa, i
      double precision ip(lia), iv(lia), ivs(lia), oa(loa)

      do 10 i=n+1,lia

        oa(i) = ( (ivs(i) - iv(i)) * oa(i-1) + iv(i) * ip(i) ) / ivs(i)

   10 continue
      end

c-----------------------------------------------------------------------c
c
c     Calculate Zero Lag Exponential Moving Average (ZLEMA)
c
c http://www.fmlabs.com/reference/ZeroLagExpMA.htm
c http://linnsoft.com/tour/techind/movAvg.htm
c
c     The above sites do not account for the possibility of 'lag' being
c     non-integer.  I implement a weighted average of the two nearest
c     'ia' values if 'lag' is non-integer.
c
c     ia    : input array
c     lia   : length of input array
c     n     : number of periods
c     oa    : output array
c     loa   : length of output array
c     ratio : weighting/decay ratio
c
      subroutine zlema(ia, lia, n, oa, loa, ratio)

      integer lia, n, loa, i, loc 
      double precision ia(lia), oa(loa)
      double precision ratio, lag, wt

      lag = 1/ratio
      wt = DMOD( lag, 1.0D0 )

      do 10 i=n+1,lia

        loc = DINT(i-(lag))
        oa(i) = ratio * ( 2*ia(i) - (ia(loc)*(1-wt) + ia(loc+1)*wt ) )
     C         + ( 1 - ratio ) * oa(i-1)

   10 continue
      end
